From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Hannes Winkler <hanneswinkler2000@web.de>
Date: Thu, 9 Nov 2023 15:04:30 +0100
Subject: [PATCH 3/8] support cross-building linux engine from macos

engine/src/build/config/BUILDCONFIG.gn:
Use mac toolchain when building for linux

engine/src/build/toolchain/linux/BUILD.gn:
Use //buildtools/mac-x64/clang... when running on mac

gn: make linux arm64 build not be auto-detected as a host build
linux embedder: always depend on `libflutter_linux_gtk.so` instead of
`libflutter_${host_os}_gtk.so`
---
 engine/src/build/config/BUILDCONFIG.gn           | 14 +++++++++-----
 engine/src/build/toolchain/linux/BUILD.gn        | 14 +++++++++-----
 engine/src/flutter/shell/platform/linux/BUILD.gn |  4 ++--
 engine/src/flutter/tools/gn                      |  4 ++--
 4 files changed, 22 insertions(+), 14 deletions(-)

diff --git a/engine/src/build/config/BUILDCONFIG.gn b/engine/src/build/config/BUILDCONFIG.gn
index 2b0d94e1..ba687ba7 100644
--- a/engine/src/build/config/BUILDCONFIG.gn
+++ b/engine/src/build/config/BUILDCONFIG.gn
@@ -586,16 +586,20 @@ if (custom_toolchain != "") {
     set_default_toolchain("//build/toolchain/android:$current_cpu")
   }
 } else if (is_linux) {
+  if (host_os == "linux") {
+    if (is_clang) {
+      host_toolchain = "//build/toolchain/linux:clang_$host_cpu"
+    } else {
+      host_toolchain = "//build/toolchain/linux:$host_cpu"
+    }
+  } else if (host_os == "mac") {
+    host_toolchain = "//build/toolchain/mac:clang_$host_cpu"
+  }
   if (is_clang) {
-    host_toolchain = "//build/toolchain/linux:clang_$host_cpu"
     set_default_toolchain("//build/toolchain/linux:clang_$current_cpu")
   } else {
-    host_toolchain = "//build/toolchain/linux:$host_cpu"
     set_default_toolchain("//build/toolchain/linux:$current_cpu")
   }
-  if (is_chromeos && cros_use_custom_toolchain) {
-    set_default_toolchain("//build/toolchain/cros:target")
-  }
 } else if (is_mac) {
   host_toolchain = "//build/toolchain/mac:clang_$host_cpu"
   set_default_toolchain("//build/toolchain/mac:clang_$current_cpu")
diff --git a/engine/src/build/toolchain/linux/BUILD.gn b/engine/src/build/toolchain/linux/BUILD.gn
index b6db422f..efbd0512 100644
--- a/engine/src/build/toolchain/linux/BUILD.gn
+++ b/engine/src/build/toolchain/linux/BUILD.gn
@@ -33,14 +33,18 @@ if (use_rbe) {
   link_prefix = ""
 }
 
-if (host_cpu == "arm64") {
-  rebased_clang_dir =
-      rebase_path("$buildtools_path/linux-arm64/clang/bin", root_build_dir)
+if (host_os == "linux" && host_cpu == "x64") {
+  host_dir = "linux-x64"
+} else if (host_os == "linux" && host_cpu == "arm64") {
+  host_dir = "linux-arm64"
+} else if (host_os == "mac" && host_cpu == "x64") {
+  host_dir = "mac-x64"
 } else {
-  rebased_clang_dir =
-      rebase_path("$buildtools_path/linux-x64/clang/bin", root_build_dir)
+  assert(false, "Unsupported host OS & CPU combination: $host_os, $host_cpu")
 }
 
+rebased_clang_dir = rebase_path("$buildtools_path/$host_dir/clang/bin", root_build_dir)
+
 gcc_toolchain("arm") {
   prefix = "arm-linux-gnueabihf-"
   if (toolchain_prefix != "") {
diff --git a/engine/src/flutter/shell/platform/linux/BUILD.gn b/engine/src/flutter/shell/platform/linux/BUILD.gn
index d64d7601..86f3bc40 100644
--- a/engine/src/flutter/shell/platform/linux/BUILD.gn
+++ b/engine/src/flutter/shell/platform/linux/BUILD.gn
@@ -345,8 +345,8 @@ zip_bundle("flutter_gtk") {
   }
   tmp_files += [
     {
-      source = "$root_build_dir/libflutter_${host_os}_gtk.so"
-      destination = "libflutter_${host_os}_gtk.so"
+      source = "$root_build_dir/libflutter_linux_gtk.so"
+      destination = "libflutter_linux_gtk.so"
     },
     {
       source = "$root_build_dir/gen_snapshot"
diff --git a/engine/src/flutter/tools/gn b/engine/src/flutter/tools/gn
index 726f1715..b09708e2 100755
--- a/engine/src/flutter/tools/gn
+++ b/engine/src/flutter/tools/gn
@@ -94,8 +94,8 @@ def is_host_build(args):
     return True
   # For linux arm64 builds, we cross compile from x64 hosts, so the
   # target_os='linux' and linux-cpu='arm64'
-  if args.target_os == 'linux' and args.linux_cpu == 'arm64':
-    return True
+  # if args.target_os == 'linux' and args.linux_cpu == 'arm64':
+  #   return True
   # The Mac and host targets are redundant. Again, necessary to disambiguate
   # during cross-compilation.
   if args.target_os == 'mac':
